pkg/api/helpers/upload.go

```diff
@@ -0,0 +1,108 @@
+func ClientUpload(ctx context.Context, client api.ClientWithResponsesInterface, repoID, branchID, filePath string, metadata map[string]string, contents io.ReadSeeker) (*api.ObjectStats, error) {
+	resp, err := client.GetPhysicalAddressWithResponse(ctx, repoID, branchID, &api.GetPhysicalAddressParams{
+		Path: filePath,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("get physical address to upload object: %w", err)
+	}
+	if resp.JSONDefault != nil {
+		return nil, fmt.Errorf("%w: %s", ErrRequestFailed, resp.JSONDefault.Message)
+	}
+	if resp.StatusCode() != http.StatusOK {
+		return nil, fmt.Errorf("%w: %s (status code %d)", ErrRequestFailed, resp.Status(), resp.StatusCode())
+	}
+
+	stagingLocation := *resp.JSON200
+	for { // Return from inside loop
+		physicalAddress := api.StringValue(stagingLocation.PhysicalAddress)
+		parsedAddress, err := url.Parse(physicalAddress)
+		if err != nil {
+			return nil, fmt.Errorf("parse physical address URL %s: %w", physicalAddress, err)
+		}
+
+		// TODO(ariels): plug-in support for other protocols
+		if parsedAddress.Scheme != "s3" {
+			return nil, fmt.Errorf("%w %s", ErrUnsupportedProtocol, parsedAddress.Scheme)
+		}
+
+		bucket := parsedAddress.Hostname()
+		sess, err := session.NewSessionWithOptions(session.Options{
+			SharedConfigState: session.SharedConfigEnable,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("connect to S3 session: %w", err)
+		}
+		sess.ClientConfig(s3.ServiceName)
+		svc := s3.New(sess)
+
+		// TODO(ariels): Allow customization of request
+		putObjectResponse, err := svc.PutObject(&s3.PutObjectInput{
+			Body:   contents,
+			Bucket: &bucket,
+			Key:    &parsedAddress.Path,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("upload to backing store %v: %w", parsedAddress, err)
+		}
+
+		size, err := contents.Seek(0, io.SeekEnd)
+		if err != nil {
+			return nil, fmt.Errorf("read size: %w", err)
+		}
+		_, err = contents.Seek(0, io.SeekStart)
+		if err != nil {
+			return nil, fmt.Errorf("rewind: %w", err)
+		}
+
+		resp, err := client.LinkPhysicalAddressWithResponse(ctx, repoID, branchID, &api.LinkPhysicalAddressParams{
+			Path: filePath,
+		}, api.LinkPhysicalAddressJSONRequestBody{
+			Checksum:  api.StringValue(putObjectResponse.ETag),
+			SizeBytes: size,
+			Staging:   stagingLocation,
+			UserMetadata: &api.StagingMetadata_UserMetadata{
+				AdditionalProperties: metadata,
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("link object to backing store: %w", err)
+		}
+		if resp.StatusCode() == http.StatusOK {
+			return &api.ObjectStats{
+				Checksum: *putObjectResponse.ETag,
+				// BUG(ariels): Unavailable on S3, remove this field entirely
+				//     OR add it to the server staging manager API.
+				Mtime:           time.Now().Unix(),
+				Path:            filePath,
+				PathType:        "object",
+				PhysicalAddress: physicalAddress,
+				SizeBytes:       &size,
+			}, nil
+		}
+		if resp.JSON409 == nil {
```

Comment:
This is confusing, but maybe it's just me.
First if checks for the status code, second if checks for the value that comes with the specific status code, without checking the status code first?

Reply:
Yes, AFAIU.  The code gen creates a response type with a separate field for each possible status code.  That field will be `nil` if the status code is different.  The converse need not be true, so above we _do_ test the status code because we ignore all response fields.
@nopcoder to confirm.